index.js

```diff
@@ -3232,6 +3232,49 @@
   S.range =
   def('range', {}, [$.Integer, $.Integer, $.Array($.Integer)], range);
 
+  //# groupBy :: (a -> a -> Boolean) -> Array a -> Array (Array a)
+  //.
+  //. Splits its array argument into an array of arrays of equal,
+  //. adjacent elements. The equality of the elements is tested with
+  //. a curried function provided by the user in the first argument.
+  //.
+  //. The behaviour of groupBy can be surprising for functions that
+  //. aren't reflexive, transitive and symmetric
+  //. (see [equivalence][] relation).
+  //.
+  //. ```javascript
+  //. > S.groupBy(S.equals, [1, 1, 2, 3, 3, 4])
+  //. [[1, 1], [2], [3, 3], [4]]
+  //.
+  //. > S.groupBy(S.equals, [1, 1, 2, 1, 1])
+  //. [[1, 1], [2], [1, 1]]
+  //.
+  //. > S.groupBy(a => b => a * b % 3 === 0, [1, 2, 3, 4, 5, 6, 7, 8, 9])
+  //. [[1], [2, 3], [4], [5, 6], [7], [8, 9]]
+  //. ```
+  function groupBy(fn, xs) {
+    if (xs.length === 0) {
+      return [];
+    }
+    var acc = [[xs[0]]];
+    for (var i = 1; i < xs.length; i += 1) {
+      var last = acc[acc.length - 1];
+      var a = last[0];
+      var b = xs[i];
+      if (fn(a)(b)) {
+        last.push(b);
+      } else {
+        acc.push([b]);
+      }
+    }
+    return acc;
+  }
```

Comment:
I also had a go. Optimized by taking out `.length` calls and random access on every iteration:

```js
function groupBy(f, xs) {
  var len = xs.length;
  if (len === 0) return [];
  var last = [xs[0]], acc = [last];
  for (var idx = 1; idx < len; idx++) {
    var a = last[0];
    var b = xs[idx];
    if (f(a)(b)) last.push(b);
    else acc.push(last = [b]);
  }
  return acc;
}
```